router/src/batcher.rs

```diff
@@ -105,7 +105,7 @@ async fn batching_task(
         // This batch might be smaller than the maximum batch size if there are not enough requests
         // waiting in the DB
         let mut waiting_tokens = 0;
-        if let Some((request_ids, batch)) = db.next_batch(None, max_batch_size) {
```

Comment:
I don't think we should change this line as it could lead to a wrong value for `waiting_tokens`.

If I understood correctly, the potential issue arises if a request comes while we are finishing a batch. The `notify_waiters` call happens while we are finishing a generation and the `batching_task` gets stuck till a new request happens and it can finally cross the `notified.await`.

Is my understanding correct? If so, what do you think of just changing [this line](https://github.com/huggingface/text-generation-inference/blob/main/router/src/batcher.rs#L73)?

```diff
- self.shared.batching_task.notify_waiters();
+ self.shared.batching_task.notify_one();
```

Reply:
> I don't think we should change this line as it could lead to a wrong value for `waiting_tokens`.

I think this could be resolved by moving the `moving_tokens` assignment after the `next_batch` call?

> Is my understanding correct?

Yes, that's correct.

> If so, what do you think of just changing [this line](https://github.com/huggingface/text-generation-inference/blob/main/router/src/batcher.rs#L73)?
> 
> ```diff
> - self.shared.batching_task.notify_waiters();
> + self.shared.batching_task.notify_one();
> ```

I think you're right that this is also necessary (I hadn't realized the differing semantics of `notify_waiited` and `notify_one`), but I don't think it's sufficient.

Consider a case where a burst of requests (> max batch size) arrives while a batch is in progress. The current batch all finish at the same time and so `None` is returned from `generate_with_cache` and we return to the start of the loop.

At this point it would be stuck at `notify_waiters`, but if this is changed to `notify_one` then we would still have a permit and so proceed to fill a new batch. However if this next max-sized batch also finished together without further requests arriving, the loop would now be stuck at `notify_one` despite requests remaining in the db queue.

Reply:
@OlivierDehaene I pushed a couple more commits addressing the above, let me know what you think. Also noticed an incorrect error message.

Reply:
Yes it makes perfect sense! 

Then I think we should keep `notify_waiters`as is as I think it is more explicit. Our intent is to notify the batching task that we have new requests but it is its responsability to empty the db with this new looping logic.

Great catch btw! It's good to have a new pair of eyes on this code :)

Reply:
@OlivierDehaene thanks, and thanks again for sharing it, I really like how clean and well-structured it is. And it got me to finally learn rust :)